File: legacy-runtime/src/main/java/com/xiaomi/channel/commonutils/misc/JobMutualExclusor.java
package com.xiaomi.channel.commonutils.misc;

import android.content.Context;
import android.os.Build;
import android.os.Environment;
import android.text.TextUtils;
import com.xiaomi.channel.commonutils.android.AppInfoUtils;
import com.xiaomi.channel.commonutils.android.DeviceInfo;
import com.xiaomi.channel.commonutils.file.IOUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.channels.FileLock;
import java.util.ArrayList;
import java.util.List;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/channel/commonutils/misc/JobMutualExclusor.class */
public class JobMutualExclusor {
    public static final float CHECK_FRACTION = 0.9f;
    private static final String INFO_SEPARATOR = ",";
    private static final String LAST_COLLECT_FILE_PATH = "lcfp";
    private static final String LAST_COLLECT_FILE_PATH_LOCK = "lcfp.lock";
    private static final String WRITE_FILE_PERMISSION = "android.permission.WRITE_EXTERNAL_STORAGE";

    public static boolean checkPeriodAndRecordWithFileLock(Context context, String str, long j) {
        if (Build.VERSION.SDK_INT >= 23 && !AppInfoUtils.checkSelfPermission(context, "android.permission.WRITE_EXTERNAL_STORAGE")) {
            return true;
        }
        RandomAccessFile randomAccessFile = null;
        FileLock fileLock = null;
        RandomAccessFile randomAccessFile2 = null;
        FileLock fileLock2 = null;
        try {
            try {
                File file = new File(new File(Environment.getExternalStorageDirectory(), DeviceInfo.VIRTUAL_DEVICE_DIR), LAST_COLLECT_FILE_PATH_LOCK);
                IOUtils.createFileQuietly(file);
                RandomAccessFile randomAccessFile3 = new RandomAccessFile(file, "rw");
                FileLock fileLockLock = randomAccessFile3.getChannel().lock();
                randomAccessFile = randomAccessFile3;
                fileLock = fileLockLock;
                randomAccessFile2 = randomAccessFile3;
                fileLock2 = fileLockLock;
                boolean zCheckPeriodAndRecordWorking = checkPeriodAndRecordWorking(context, str, j);
                if (fileLockLock != null && fileLockLock.isValid()) {
                    try {
                        fileLockLock.release();
                    } catch (IOException e) {
                    }
                }
                IOUtils.closeQuietly(randomAccessFile3);
                return zCheckPeriodAndRecordWorking;
            } catch (IOException e2) {
                randomAccessFile = randomAccessFile2;
                fileLock = fileLock2;
                e2.printStackTrace();
                if (fileLock2 != null && fileLock2.isValid()) {
                    try {
                        fileLock2.release();
                    } catch (IOException e3) {
                    }
                }
                IOUtils.closeQuietly(randomAccessFile2);
                return true;
            }
        } catch (Throwable th) {
            if (fileLock != null && fileLock.isValid()) {
                try {
                    fileLock.release();
                } catch (IOException e4) {
                }
            }
            IOUtils.closeQuietly(randomAccessFile);
            if (th instanceof RuntimeException) {
                throw (RuntimeException) th;
            }
            throw new RuntimeException(th);
        }
    }

    private static boolean checkPeriodAndRecordWorking(Context context, String str, long j) throws Throwable {
        if (TextUtils.isEmpty(str) || j <= 0) {
            return true;
        }
        File file = new File(new File(Environment.getExternalStorageDirectory(), DeviceInfo.VIRTUAL_DEVICE_DIR), LAST_COLLECT_FILE_PATH);
        IOUtils.createFileQuietly(file);
        String strFileToStr = IOUtils.fileToStr(file);
        long jCurrentTimeMillis = System.currentTimeMillis();
        long j2 = (long) (j * CHECK_FRACTION);
        String packageName = context.getPackageName();
        List<String> arrayList = new ArrayList<>();
        boolean z = true;
        boolean z2 = false;
        if (!TextUtils.isEmpty(strFileToStr)) {
            String[] strArrSplit = strFileToStr.split("\n");
            for (String str2 : strArrSplit) {
                if (TextUtils.isEmpty(str2)) {
                    continue;
                }
                String[] strArrSplit2 = str2.split(INFO_SEPARATOR);
                if (strArrSplit2.length < 3) {
                    continue;
                }
                if (TextUtils.equals(str, strArrSplit2[0])) {
                    z2 = true;
                    try {
                        long j3 = Long.parseLong(strArrSplit2[2]);
                        if (Math.abs(jCurrentTimeMillis - j3) < j2) {
                            z = false;
                        }
                    } catch (NumberFormatException e) {
                        MyLog.e(e);
                    }
                    if (z) {
                        arrayList.add(str + INFO_SEPARATOR + packageName + INFO_SEPARATOR + jCurrentTimeMillis);
                    } else {
                        arrayList.add(str2);
                    }
                } else {
                    arrayList.add(str2);
                }
            }
        }
        if (z && !z2) {
            arrayList.add(str + INFO_SEPARATOR + packageName + INFO_SEPARATOR + jCurrentTimeMillis);
        }
        if (z) {
            StringBuilder sb = new StringBuilder();
            for (String str3 : arrayList) {
                if (!TextUtils.isEmpty(str3)) {
                    if (sb.length() > 0) {
                        sb.append('\n');
                    }
                    sb.append(str3);
                }
            }
            IOUtils.strToFile(file, sb.toString());
        }
        return z;
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/channel/commonutils/misc/JobMutualExclusor.kt
package com.xiaomi.channel.commonutils.misc

import android.content.Context
import android.os.Build
import android.os.Environment
import android.text.TextUtils
import com.xiaomi.channel.commonutils.android.AppInfoUtils
import com.xiaomi.channel.commonutils.android.DeviceInfo
import com.xiaomi.channel.commonutils.file.IOUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.channels.FileLock
import kotlin.math.abs

/*
 * Current override reference: com.xiaomi.xmsf 0.3.17-20260410000745 (versionCode 1003003000),
 * base.apk sha256 f3d72b6f5e1427ceecd3147a051d58e4dc95bb528397d486658e01cad9f7e590,
 * JADX path: com.xiaomi.xmsf/current/base/sources/com/xiaomi/channel/commonutils/misc/JobMutualExclusor.java
 */
object JobMutualExclusor {
    const val CHECK_FRACTION = 0.9f
    private const val INFO_SEPARATOR = ","
    private const val LAST_COLLECT_FILE_PATH = "lcfp"
    private const val LAST_COLLECT_FILE_PATH_LOCK = "lcfp.lock"
    private const val WRITE_FILE_PERMISSION = "android.permission.WRITE_EXTERNAL_STORAGE"

    @JvmStatic
    fun checkPeriodAndRecordWithFileLock(context: Context, str: String?, period: Long): Boolean {
        if (Build.VERSION.SDK_INT >= 23 && !AppInfoUtils.checkSelfPermission(context, WRITE_FILE_PERMISSION)) {
            return true
        }
        var randomAccessFile: RandomAccessFile? = null
        var fileLock: FileLock? = null
        return try {
            val file = File(File(Environment.getExternalStorageDirectory(), DeviceInfo.VIRTUAL_DEVICE_DIR), LAST_COLLECT_FILE_PATH_LOCK)
            IOUtils.createFileQuietly(file)
            randomAccessFile = RandomAccessFile(file, "rw")
            fileLock = randomAccessFile.channel.lock()
            checkPeriodAndRecordWorking(context, str, period)
        } catch (e: IOException) {
            e.printStackTrace()
            true
        } finally {
            if (fileLock != null && fileLock!!.isValid) {
                try {
                    fileLock!!.release()
                } catch (e: IOException) {
                    // ignore
                }
            }
            IOUtils.closeQuietly(randomAccessFile)
        }
    }

    private fun checkPeriodAndRecordWorking(context: Context, str: String?, period: Long): Boolean {
        if (TextUtils.isEmpty(str) || period <= 0) {
            return true
        }
        val file = File(File(Environment.getExternalStorageDirectory(), DeviceInfo.VIRTUAL_DEVICE_DIR), LAST_COLLECT_FILE_PATH)
        IOUtils.createFileQuietly(file)
        val existing = IOUtils.fileToStr(file)
        val now = System.currentTimeMillis()
        val threshold = (period * CHECK_FRACTION).toLong()
        val packageName = context.packageName
        val lines = ArrayList<String>()
        var allowed = true
        var found = false
        if (!TextUtils.isEmpty(existing)) {
            for (line in existing!!.split("\n")) {
                if (TextUtils.isEmpty(line)) {
                    continue
                }
                val parts = line.split(INFO_SEPARATOR)
                if (parts.size < 3) {
                    continue
                }
                if (TextUtils.equals(str, parts[0])) {
                    found = true
                    try {
                        val last = parts[2].toLong()
                        if (abs(now - last) < threshold) {
                            allowed = false
                        }
                    } catch (e: NumberFormatException) {
                        MyLog.e(e)
                    }
                    if (allowed) {
                        lines.add(str + INFO_SEPARATOR + packageName + INFO_SEPARATOR + now)
                    } else {
                        lines.add(line)
                    }
                } else {
                    lines.add(line)
                }
            }
        }
        if (allowed && !found) {
            lines.add(str + INFO_SEPARATOR + packageName + INFO_SEPARATOR + now)
        }
        if (allowed) {
            IOUtils.strToFile(file, lines.filter { !TextUtils.isEmpty(it) }.joinToString("\n"))
        }
        return allowed
    }
}
